components/odh-notebook-controller/controllers/notebook_webhook.go

```diff
@@ -438,3 +452,112 @@ func InjectCertConfig(notebook *nbv1.Notebook, configMapName string) error {
 	}
 	return nil
 }
+
+// This function checks if there is an internal registry and takes the corresponding actions to set the container.image value.
+// If an internal registry is detected, it uses the default values specified in the Notebook Custom Resource (CR).
+// Otherwise, it checks the last-image-selection annotation to find the image stream and fetches the image from status.dockerImageReference,
+// assigning it to the container.image value.
+func SetContainerImageFromRegistry(ctx context.Context, cli client.Client, notebook *nbv1.Notebook, log logr.Logger) error {
+
+	// Load kubeconfig
```

Comment:
Here's another way how to get a kubeconfig. Get the config that's in use passed in as parameter.

```diff
diff --git a/components/odh-notebook-controller/controllers/notebook_webhook.go b/components/odh-notebook-controller/controllers/notebook_webhook.go
index 0a94bd5a..74a27862 100644
--- a/components/odh-notebook-controller/controllers/notebook_webhook.go
+++ b/components/odh-notebook-controller/controllers/notebook_webhook.go
@@ -20,8 +20,6 @@ import (
 	"encoding/json"
 	"fmt"
 	"net/http"
-	"os"
-	"path/filepath"
 	"sort"
 	"strings"
 
@@ -36,7 +34,6 @@ import (
 	"k8s.io/apimachinery/pkg/util/intstr"
 	"k8s.io/client-go/dynamic"
 	"k8s.io/client-go/rest"
-	"k8s.io/client-go/tools/clientcmd"
 	"k8s.io/utils/pointer"
 	"sigs.k8s.io/controller-runtime/pkg/client"
 	"sigs.k8s.io/controller-runtime/pkg/webhook/admission"
@@ -48,6 +45,7 @@ import (
 type NotebookWebhook struct {
 	Log         logr.Logger
 	Client      client.Client
+	Config      *rest.Config
 	Decoder     *admission.Decoder
 	OAuthConfig OAuthConfig
 }
@@ -256,7 +254,7 @@ func (w *NotebookWebhook) Handle(ctx context.Context, req admission.Request) adm
 		}
 
 		// Check Imagestream Info
-		err = SetContainerImageFromRegistry(ctx, w.Client, notebook, log)
+		err = SetContainerImageFromRegistry(ctx, w.Config, notebook, log)
 		if err != nil {
 			return admission.Errored(http.StatusInternalServerError, err)
 		}
@@ -457,24 +455,9 @@ func InjectCertConfig(notebook *nbv1.Notebook, configMapName string) error {
 // If an internal registry is detected, it uses the default values specified in the Notebook Custom Resource (CR).
 // Otherwise, it checks the last-image-selection annotation to find the image stream and fetches the image from status.dockerImageReference,
 // assigning it to the container.image value.
-func SetContainerImageFromRegistry(ctx context.Context, cli client.Client, notebook *nbv1.Notebook, log logr.Logger) error {
-
-	// Load kubeconfig
-	config, err := rest.InClusterConfig()
-	if err != nil {
-		kubeconfig := os.Getenv("KUBECONFIG")
-		if kubeconfig == "" {
-			kubeconfig = filepath.Join(os.Getenv("HOME"), ".kube", "config")
-		}
-		config, err = clientcmd.BuildConfigFromFlags("", kubeconfig)
-		if err != nil {
-			log.Error(err, "Error creating config")
-			return err
-		}
-	}
-
+func SetContainerImageFromRegistry(ctx context.Context, cli *rest.Config, notebook *nbv1.Notebook, log logr.Logger) error {
 	// Create a dynamic client
-	dynamicClient, err := dynamic.NewForConfig(config)
+	dynamicClient, err := dynamic.NewForConfig(cli)
 	if err != nil {
 		log.Error(err, "Error creating dynamic client")
 		return err
diff --git a/components/odh-notebook-controller/controllers/suite_test.go b/components/odh-notebook-controller/controllers/suite_test.go
index 8ffc9557..4cbc05cc 100644
--- a/components/odh-notebook-controller/controllers/suite_test.go
+++ b/components/odh-notebook-controller/controllers/suite_test.go
@@ -18,9 +18,7 @@ import (
 	"context"
 	"crypto/tls"
 	"fmt"
-	"io/ioutil"
 	"net"
-	"os"
 	"path/filepath"
 	"testing"
 	"time"
@@ -84,37 +82,7 @@ var _ = BeforeSuite(func() {
 	}
 	logf.SetLogger(zap.New(zap.WriteTo(GinkgoWriter), zap.UseFlagOptions(&opts)))
 
-	// Create a temporary kubeconfig file
-	kubeconfigContent := []byte(`
-apiVersion: v1
-clusters:
-- cluster:
-    server: https://localhost:6443
-  name: local
-contexts:
-- context:
-    cluster: local
-    user: user
-  name: local
-current-context: local
-kind: Config
-preferences: {}
-users:
-- name: user
-  user:
-    token: fake-token
-`)
-	tmpDir, err := ioutil.TempDir("", "kubeconfig")
-	Expect(err).NotTo(HaveOccurred())
-
-	kubeconfigPath := filepath.Join(tmpDir, "config")
-	err = ioutil.WriteFile(kubeconfigPath, kubeconfigContent, 0644)
-	Expect(err).NotTo(HaveOccurred())
-
-	// Set the KUBECONFIG environment variable
-	os.Setenv("KUBECONFIG", kubeconfigPath)
-
-	// Initiliaze test environment:
+	// Initialize test environment:
 	// https://pkg.go.dev/sigs.k8s.io/controller-runtime/pkg/envtest#Environment.Start
 	By("Bootstrapping test environment")
 	envTest = &envtest.Environment{
@@ -129,7 +97,7 @@ users:
 		},
 	}
 
-	cfg, err = envTest.Start()
+	cfg, err := envTest.Start()
 	Expect(err).NotTo(HaveOccurred())
 	Expect(cfg).NotTo(BeNil())
 
@@ -172,6 +140,7 @@ users:
 		Handler: &NotebookWebhook{
 			Log:    ctrl.Log.WithName("controllers").WithName("notebook-controller"),
 			Client: mgr.GetClient(),
+			Config: mgr.GetConfig(),
 			OAuthConfig: OAuthConfig{
 				ProxyImage: OAuthProxyImage,
 			},
diff --git a/components/odh-notebook-controller/main.go b/components/odh-notebook-controller/main.go
index f4adc8b0..bef4615c 100644
--- a/components/odh-notebook-controller/main.go
+++ b/components/odh-notebook-controller/main.go
@@ -116,6 +116,7 @@ func main() {
 		Handler: &controllers.NotebookWebhook{
 			Log:    ctrl.Log.WithName("controllers").WithName("Notebook"),
 			Client: mgr.GetClient(),
+			Config: mgr.GetConfig(),
 			OAuthConfig: controllers.OAuthConfig{
 				ProxyImage: oauthProxyImage,
 			},
```